lib/Sema/ConstraintSystem.cpp

```diff
@@ -2022,7 +2022,7 @@ TypeVariableType *ConstraintSystem::openGenericParameter(
   if (parameter->isParameterPack())
     options |= TVO_CanBindToPack;
 
-  if (shouldAttemptFixes())
+  else if (shouldAttemptFixes())
     options |= TVO_CanBindToHole;
```

Comment:
Let's do this slightly differently so it a bit more readable:

```
if (shouldAttemptFixes())
    options |= TVO_CanBindToHole;

  if (parameter->isParameterPack()) {
    options |= TVO_CanBindToPack;
    options -= TVO_CanBindToHole;
  }
```

Reply:
Hi! Just pushed the suggested version. But I found this is not fixed:

```swift
struct S<each T> {
    init(_ values: repeat each T) {}
}

let _ = S.init(1, nil) // OK, Correct diagnosis: 'nil' requires a contextual type
let _ = S.init(1, nil) // OK, Correct diagnosis: 'nil' requires a contextual type
let _ = S(1, nil) // OK, Correct diagnosis: 'nil' requires a contextual type
let _ = S(nil) // ❌
```

Any idea on this?

```
Assertion failed: (!fixed->is<TypeVariableType>() && "Cannot bind to type variable; merge equivalence classes instead"), function bindTypeVariable, file ConstraintGraph.cpp, line 605.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend /Users/lizhen/Projects/SwiftPlayground/Sources/main.swift -typecheck
1.	Swift version 5.11-dev (LLVM c4000d8995e87b5, Swift 6fdfd7ea81a5c12)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "/Users/lizhen/Projects/SwiftPlayground/Sources/main.swift")
4.	While type-checking statement at [/Users/lizhen/Projects/SwiftPlayground/Sources/main.swift:8:1 - line:8:14] RangeText="let a = A(nil"
5.	While type-checking declaration 0x13d19b8f8 (at /Users/lizhen/Projects/SwiftPlayground/Sources/main.swift:8:1)
6.	While evaluating request PatternBindingEntryRequest((unknown decl), 0)
7.	While type-checking expression at [/Users/lizhen/Projects/SwiftPlayground/Sources/main.swift:8:9 - line:8:14] RangeText="A(nil"
8.	While type-checking-target starting at /Users/lizhen/Projects/SwiftPlayground/Sources/main.swift:8:9
 #0 0x000000010832ea38 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x105536a38)
 #1 0x000000010832d228 llvm::sys::RunSignalHandlers() (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x105535228)
 #2 0x000000010832f080 SignalHandler(int) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x105537080)
 #3 0x0000000184755a24 (/usr/lib/system/libsystem_platform.dylib+0x18046da24)
 #4 0x0000000184725cc0 (/usr/lib/system/libsystem_pthread.dylib+0x18043dcc0)
 #5 0x0000000184631a40 (/usr/lib/system/libsystem_c.dylib+0x180349a40)
 #6 0x0000000184630d30 (/usr/lib/system/libsystem_c.dylib+0x180348d30)
 #7 0x0000000108711b98 swift::constraints::ConstraintGraph::bindTypeVariable(swift::TypeVariableType*, swift::Type) (.cold.4) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x105919b98)
 #8 0x00000001040defc8 swift::constraints::ConstraintGraph::bindTypeVariable(swift::TypeVariableType*, swift::Type) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1012e6fc8)
 #9 0x00000001040f126c swift::constraints::ConstraintSystem::assignFixedType(swift::TypeVariableType*, swift::Type, bool, bool) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1012f926c)
#10 0x0000000104030ef0 swift::constraints::ConstraintSystem::matchTypesBindTypeVar(swift::TypeVariableType*, swift::Type, swift::constraints::ConstraintKind, swift::optionset::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder, llvm::function_ref<swift::constraints::ConstraintSystem::TypeMatchResult ()>) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x101238ef0)
#11 0x000000010402669c swift::constraints::ConstraintSystem::matchTypes(swift::Type, swift::Type, swift::constraints::ConstraintKind, swift::optionset::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x10122e69c)
#12 0x0000000104050484 swift::constraints::ConstraintSystem::simplifyConstraint(swift::constraints::Constraint const&) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x101258484)
#13 0x0000000104064f74 swift::constraints::ConstraintSystem::simplify() (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x10126cf74)
#14 0x0000000103fee970 swift::constraints::TypeVariableBinding::attempt(swift::constraints::ConstraintSystem&) const (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1011f6970)
#15 0x000000010407b2fc swift::constraints::BindingStep<swift::constraints::TypeVarBindingProducer>::take(bool) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1012832fc)
#16 0x0000000104067090 swift::constraints::ConstraintSystem::solveImpl(llvm::SmallVectorImpl<swift::constraints::Solution>&) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x10126f090)
#17 0x00000001040ff3b0 swift::constraints::ConstraintSystem::salvage() (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1013073b0)
#18 0x0000000104068610 swift::constraints::ConstraintSystem::solve(swift::constraints::SyntacticElementTarget&, swift::FreeTypeVariableBinding) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x101270610)
#19 0x00000001041e3c7c swift::TypeChecker::typeCheckTarget(swift::constraints::SyntacticElementTarget&, swift::optionset::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1013ebc7c)
#20 0x00000001041e3afc swift::TypeChecker::typeCheckExpression(swift::constraints::SyntacticElementTarget&, swift::optionset::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1013ebafc)
#21 0x00000001041e4bf8 swift::TypeChecker::typeCheckBinding(swift::Pattern*&, swift::Expr*&, swift::DeclContext*, swift::Type, swift::PatternBindingDecl*, unsigned int, swift::optionset::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1013ecbf8)
#22 0x00000001041e4e2c swift::TypeChecker::typeCheckPatternBinding(swift::PatternBindingDecl*, unsigned int, swift::Type, swift::optionset::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1013ece2c)
#23 0x00000001042b2c3c swift::PatternBindingEntryRequest::evaluate(swift::Evaluator&, swift::PatternBindingDecl*, unsigned int) const (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014bac3c)
#24 0x00000001046a8ef4 swift::PatternBindingEntryRequest::OutputType swift::Evaluator::getResultUncached<swift::PatternBindingEntryRequest, swift::PatternBindingEntryRequest::OutputType swift::evaluateOrDefault<swift::PatternBindingEntryRequest>(swift::Evaluator&, swift::PatternBindingEntryRequest, swift::PatternBindingEntryRequest::OutputType)::'lambda'()>(swift::PatternBindingEntryRequest const&, swift::PatternBindingEntryRequest::OutputType swift::evaluateOrDefault<swift::PatternBindingEntryRequest>(swift::Evaluator&, swift::PatternBindingEntryRequest, swift::PatternBindingEntryRequest::OutputType)::'lambda'()) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1018b0ef4)
#25 0x000000010467767c swift::PatternBindingDecl::getCheckedPatternBindingEntry(unsigned int) const (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x10187f67c)
#26 0x0000000104229044 (anonymous namespace)::DeclChecker::visit(swift::Decl*) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x101431044)
#27 0x0000000104228bac swift::TypeChecker::typeCheckDecl(swift::Decl*) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x101430bac)
#28 0x00000001042aa44c swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014b244c)
#29 0x00000001042a8a90 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014b0a90)
#30 0x00000001042a8b40 swift::TypeChecker::typeCheckTopLevelCodeDecl(swift::TopLevelCodeDecl*) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014b0b40)
#31 0x00000001042e4df4 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014ecdf4)
#32 0x00000001042e6860 swift::TypeCheckSourceFileRequest::OutputType swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType)::'lambda'()>(swift::TypeCheckSourceFileRequest const&, swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType)::'lambda'()) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014ee860)
#33 0x00000001042e4cf0 swift::performTypeChecking(swift::SourceFile&) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1014eccf0)
#34 0x0000000103224414 bool llvm::function_ref<bool (swift::SourceFile&)>::callback_fn<swift::CompilerInstance::performSema()::$_7>(long, swift::SourceFile&) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x10042c414)
#35 0x000000010321e1b0 swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<bool (swift::SourceFile&)>) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1004261b0)
#36 0x000000010321e0f4 swift::CompilerInstance::performSema() (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1004260f4)
#37 0x0000000102ffef60 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x100206f60)
#38 0x0000000102ff08e8 performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1001f88e8)
#39 0x0000000102fef944 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1001f7944)
#40 0x0000000102e2b0d4 swift::mainEntry(int, char const**) (/Users/lizhen/Projects/swift-projects/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend+0x1000330d4)
#41 0x00000001843a50e0
```

Reply:
This looks like a bigger issue where the solver is trying to bind type variables together, most likely inference problem. I think we can land this change and fix `S(nil)` in a separate PR.

Reply:
I will take a look at the `-debug-constraints` and give you some pointers on Monday.

Reply:
Thanks! Shall I remove `S(nil)` from test cases?

Reply:
Leave the improved ones but not `S(nil)` we still want to make sure that these don’t regress :)

Reply:
Okay. But sorry that I've already broke the test with this PR 🥲. I'll try to look into this.

```swift
Command Output (stderr):
--
/Users/lizhen/Projects/swift-projects/swift/test/Constraints/variadic_generic_functions.swift:35:68: error: incorrect message found
  multipleParameters(xs: "", 5.0, ys: 5.0, "") // expected-error {{conflicting arguments to generic parameter 'each T' ('Pack{Double, String}' vs. 'Pack{String, String}' vs. 'Pack{String, Double}' vs. 'Pack{Double, Double}')}}
                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                                   conflicting arguments to generic parameter 'each T' ('Pack{Double, Double}' vs. 'Pack{Double, String}' vs. 'Pack{String, Double}' vs. 'Pack{String, String}')
/Users/lizhen/Projects/swift-projects/swift/test/Constraints/variadic_generic_functions.swift:75:60: error: expected note not produced
  func foo<each T>(_: repeat each T = bar().element) {} // expected-note {{in call to function 'foo'}}
                                                       ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/Users/lizhen/Projects/swift-projects/swift/test/Constraints/variadic_generic_functions.swift:77:6: error: expected error not produced
  // expected-error@-2 {{value pack expansion can only appear inside a function argument list, tuple element, or as the expression of a for-in loop}}
~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lizhen/Projects/swift-projects/swift/test/Constraints/variadic_generic_functions.swift:78:26: error: incorrect message found
  // expected-error@-3 {{generic parameter 'each T' could not be inferred}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         type of expression is ambiguous without a type annotation

--

********************
********************
Failed Tests (1):
  Swift(macosx-arm64) :: Constraints/variadic_generic_functions.swift
```

Reply:
Hi, I think I fixed this issue this approach:

> Other way of approaching this would be in TypeVariableBinding::fixForHole to check whether TypeVar's locator ends with PackElement and if so either:
> a. Create and record a new fix that is going to produce a diagnostic similar to type of pack expansion 'each T' cannot be inferred'.

It passes all the test under `test/Constraints` (Crash of `S(nil)` remains still)